### ISSUE TITLE
mouse_refocus = false does not work

### ISSUE THREAD

**Issue by LanderMoerkerke** [hyprwm/Hyprland#4126]
### Hyprland Version

v0.33.1

### Bug or Regression?

Bug

### Description

Following the guide on how to setup Hyprland, the setting input:mouse_refocus = false does not seem give the desirable effect.

### How to reproduce

Use config below to reproduce. Switch focus with the keyboard and see that the mouse refocuses even though the setting is turned to false.

```config
monitor=,preferred,auto,1

input {
    mouse_refocus = false
}

$mainMod = SUPER

bind = $mainMod, 1, workspace, 1
bind = $mainMod, 2, workspace, 2

bind = $mainMod, left, movefocus, l
bind = $mainMod, right, movefocus, r
bind = $mainMod, up, movefocus, u
bind = $mainMod, down, movefocus, d

bind = $mainMod, J, layoutmsg, cyclenext
bind = $mainMod, K, layoutmsg, cycleprev

bind = $mainMod, Return, exec, alacritty
```

### Crash reports, logs, images, videos

https://github.com/hyprwm/Hyprland/assets/20244592/f154afab-023f-48c0-ba90-ef8c852434d5

[hyprland-logs.txt](https://github.com/hyprwm/Hyprland/files/13640771/hyprland.txt)

**Comment by aos** [comment]
Can confirm this also happens to me. I'm on `v0.34.0`

**Comment by LRitzdorf** [comment]
I may be very wrong here, but based on (my reading of) the wiki, it seems that `mouse_refocus = false` is supposed to only switch focus when the mouse crosses a window border.

For example: with cursor warps disabled, I use my keyboard to focus window B, while leaving my mouse over window A. I would expect to be able to perform "indirect" mouse actions, like scrolling and hovering, on window A, while window B retains keyboard focus. However, once my mouse crosses over a border onto another window, that window would get keyboard focus as well.

If this is the intent, it doesn't seem to work for me, either. I can switch focus to window B using my keyboard without warping the mouse (thanks to having set `general/no_cursor_warps = true`), but the moment I move my mouse within window A (i.e. without crossing a border), window A gets keyboard focus again.

**Comment by LRitzdorf** [comment]
Not sure precisely what's changed, but this seems to have been fixed now (in v0.40.0 or earlier). The setup I describe above, with `follow_mouse = 1`, `mouse_refocus = false`, and `no_cursor_warps = true`, functions as I'd expect. Specifically:
- Window focus can be moved via the keyboard, without affecting the mouse
- Performing "indirect" mouse actions (basically anything that's not a click) on an unfocused window keeps that window unfocused
- When the mouse crosses a window border, the newly-hovered window gets focus

This lines up with my expectations from reading of the wiki.

**Comment by vaxerski** [comment]
closing then

**Comment by uwidev** [comment]
I'm a bit confused on if this is working correctly. I'm on the current commit, version `0.40.0.r117.e419ef18-1`.

The current documentation for this commit states `mouse_refocus` as follows:
> If disabled, mouse focus won’t switch to the hovered window unless the mouse crosses a window boundary when follow_mouse=1.

And with `follow_mouse=1`:
> 1 - Cursor movement will always change focus to the window under the cursor.

With `mouse_refocus` set to `false`, I interpret that it's expect that, with my keyboard, I can switch keyboard focus from my my current window A to window B. I should be able to send key inputs to keyboard scroll on B, while mouse scrolling (and perhaps move my cursor) freely on A.

In other words, this allows the user to scroll two windows simultaneously.

But currently, mouse scrolling works inconsistently. Starting on window Alacritty A, if I keyboard focus onto Floorp (Firefox) B, mouse scrolling does not work. However, it works going in reverse; that is, Starting from window (Floorp) A, keyboard switching to Alacritty B has mouse scrolling work... but on B, not on A where my mouse is at.

I feel like a potential change that could make this easier to manage is to distinctly split focus into  "mouse" and "keyboard". Making this concept clear in the wiki, and potentially redesigning how focus works could make this more clear.

Perhaps for individual mouse and keyboard focus, allow the user toggle some rules on how to switch:
1. on window border cross
2. on mouse move
3. on mouse click
4. on binding to left, right, up, down

As how things work right now, `follow_mouse=1` apply (1, 2, 3), `=2` apply (3), and `=3` applies bindings (4). What i would like to do is (1, 3, maybe 4).

And just to make things clear, mouse focus defines that mouse navigation is passed to that window (maybe clicks too somehow?). Keyboard focus is, of course, that all key inputs are passed to said window.

This discussion may warrant its own issue. I can expand further there later, but currently, `mouse_refocus` isn't working, as it seems expected to turn off rule (2) while keeping rule (1).

**Comment by LRitzdorf** [comment]
I've just performed your browser/terminal scrolling tests on my system (running v0.40.0, commit `cba1ade8`), and focusing/scrolling work as expected — i.e. scroll events are passed to the window which my cursor is over. Not to say the issue doesn't exist; it's just newer than the current Arch package.

That's a good point regarding configuration logic; I certainly found it hard to follow initially as well. Given that `mouse_refocus` only takes effect (per the docs) when `follow_mouse=1`, I feel like it might make sense to add a fifth `follow_mouse` mode which acts like `follow_mouse=1` and `mouse_refocus=false`? Just my two cents.

**Comment by uwidev** [comment]
@LRitzdorf could you share with me your `hyprland.conf` I'm currently running the most minimal setup and I cannot get any mouse scrolling to work. I also reverted back to v0.40.0 release, non-git. I don't think I mentioned it, but I run Arch.
```
monitor=,preferred,auto,1

general {
    no_cursor_warps = true
}

input {
    follow_mouse = 1
    mouse_refocus = false
}

$mainMod = SUPER

bind = $mainMod, left, movefocus, l
bind = $mainMod, right, movefocus, r
bind = $mainMod, up, movefocus, u
bind = $mainMod, down, movefocus, d

bind = $mainMod, q, exec, alacritty
bind = $mainMod, r, exec, rofi -show run
bind = $mainMod, k, killactive,
```
I open two Firefox windows, switch keyboard focus, and I actually can't get any mouse scrolling working at all. Tried with two pcmanfs-qt windows as well, and I could not get any scrolling to work.

Another thing, the moment you move your mouse, even without crossing any window boundaries, Does it bring focus back to the window the mouse moved on?

**Comment by LRitzdorf** [comment]
...okay, this is weird:
- Tested with @uwidev's minimal config, which still shows this bug (i.e. moving the cursor within a window forces that window into focus)
- Pared down my own config until the bug appeared:
  - I normally use hy3, which provides its own `hy3:movefocus` dispatcher. Once I reached the point of replacing these with the standard Hyprland `movefocus`, the bug reappeared
- Returning to my initial config, I found that replacing even a single instance of `hy3:movefocus` with the standard `movefocus` was sufficient to revive the bug

It looks like my initial "fixed" report was actually the result of my having added hy3 between my two tests. I'd assumed this wouldn't affect the cursor/window-focus interaction, but apparently I was wrong.

I have no idea why this would be — @vaxerski (apologies for the ping), any ideas? I'd be happy to test more stuff, as needed. (Also we should probably reopen this issue?)

**Comment by vaxerski** [comment]
no clue, it wasnt me who wrote the feaature and I dont use it so it's possible it was broken some day

**Comment by LRitzdorf** [comment]
As someone not familiar with C++, it looks like `CInputManager::mouseMoveUnified()` [does some conditional focusing](https://github.com/hyprwm/Hyprland/blob/255272ea18a1bd899168b2d2063b6f153edda919/src/managers/input/InputManager.cpp#L483-L492) whenever mouse movement occurs. From what I can tell, this focusing doesn't consult the `mouse_refocus` option — it seems to be purely based on various internal checks.

**Comment by df-salaz** [comment]
Can confirm that this variable is still broken. Current documentation describes this working the same as Sway's default behavior, but it does not. Top example is Hyprland.

https://github.com/user-attachments/assets/3076d538-7125-43a7-b599-f072d32f1b75

https://github.com/user-attachments/assets/78490aaa-177a-4388-90d3-88fdf3bb27d0

```
❯ hyprctl version
Hyprland, built from branch  at commit 9a09eac79b85c846e3a865a9078a3f8ff65a9259  (props: bump version to 0.42.0).
Date: Wed Aug 7 19:17:10 2024
Tag: v0.42.0, commits: 5069
```

**Comment by r3k2** [comment]
I am also having annoying problems with windows focus, I will be on vim writing code hit my mouse just a bit by mistake and loose focus from VIM !!!!  I have try mouse_refocus but is not working as expected like in "sway" 
in sway with a simple:
```
focus_follows_mouse no
mouse_warping none
```
works great! 

also is there a way to move to open a window at top, left, right or button of another window like in sway? is annoying that hyprland will just randomly select that for me and 99% of the time wrong.

Thanks

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!
